SpeechResponder/CustomFunctions/ShipCallsign.cs

```diff
@@ -17,36 +20,87 @@ public class ShipCallsign : ICustomFunction
         public string description => Properties.CustomFunctions_Untranslated.ShipCallsign;
         public NativeFunction function => new NativeFunction((values) =>
         {
-            int? localId = (values.Count == 0 ? (int?)null : (int)values[0].AsNumber);
+            // The game provides three options for callsigns used by in-game ATC:
+            // (1) CommanderName (default): ship manufacturer name (sometimes partial) + first 3 alphanumeric characters of commander name
+            // (2) ShipName: ship manufacturer name (sometimes partial) + first 3 alphanumeric characters of ship name
+            // (3) ShipID: ship manufacturer name (sometimes partial) + first 3 alphanumeric characters of ship identifier
+
+            int? localId = (values.Count < 0 && values[0].Type != ValueContent.Void 
+                ? (int)values[0].AsNumber 
+                : (int?)null);
+            int? callsignType = (values.Count > 1 && values[1].Type != ValueContent.Void 
+                ? (int)values[1].AsNumber 
+                : (int?)null);
+
             ShipMonitor shipMonitor = (ShipMonitor)EDDI.Instance.ObtainMonitor("Ship monitor");
-            Ship ship = shipMonitor.GetShip(localId);
+            Ship ship = localId is null ? shipMonitor.GetCurrentShip() : shipMonitor.GetShip(localId);
+
+            string result = "";
 
-            string result;
-            if (ship != null)
+            if (string.IsNullOrEmpty(ship?.manufacturer)) { return result; }
+
+            // First, obtain the phonetic manufacturer. This may be the complete name or only a partial name, depending on the manufacturer.
+            var phoneticmanufacturer = ShipDefinitions.ManufacturerPhoneticNames.FirstOrDefault(m => m.Key == ship.manufacturer).Value;
+
+            switch (ship.manufacturer)
             {
-                if (EDDI.Instance.Cmdr != null && EDDI.Instance.Cmdr.name != null)
-                {
-                    // Obtain the first three characters
-                    string chars = new Regex("[^a-zA-Z0-9]").Replace(EDDI.Instance.Cmdr.name, "").ToUpperInvariant().Substring(0, 3);
-                    result = ship.phoneticmanufacturer + " " + Translations.ICAO(chars);
-                }
-                else
-                {
-                    if (string.IsNullOrEmpty(ship.phoneticmanufacturer))
+                case "Core Dynamics":
```

Comment:
Would it be easier to just add a Dictionary in `ShipDefinitions.cs`, similar to `ManufacturerPhoneticNames`?

Reply:
Perhaps, but this is a rather specialized application isn't it?